src/main/scala/com/datastax/gatling/plugin/utils/CqlPreparedStatementUtil.scala

```diff
@@ -9,127 +9,209 @@ package com.datastax.gatling.plugin.utils
 
 import scala.collection.JavaConverters._
 import scala.util.matching.Regex
 
-
 trait CqlPreparedStatementUtil {
 
   protected val hourMinSecRegEx: Regex = """(\d+):(\d+):(\d+)""".r
   protected val hourMinSecNanoRegEx: Regex = """(\d+):(\d+):(\d+).(\d+{1,9})""".r
 
-  def bindParamByOrder(gatlingSession: Session,
-                       boundStatement: BoundStatement, paramType: DataType.Name,
-                       paramName: String, key: Int): BoundStatement
+  def bindParamByOrder[T <: Bindable[T]](gatlingSession: Session,
+                                         bindable: T,
+                                         paramType: DataType,
+                                         paramName: String,
+                                         key: Int): T
 
-  def bindParamByName(gatlingSession: Session, boundStatement: BoundStatement, paramType: DataType.Name,
-                      paramName: String): BoundStatement
+  def bindParamByName[T <: Bindable[T]](gatlingSession: Session,
+                                        bindable: T,
+                                        paramType: DataType,
+                                        paramName: String): T
 
-  def getParamsMap(preparedStatement: PreparedStatement): Map[String, DataType.Name]
+  def getParamsMap(preparedStatement: PreparedStatement): Map[String, DataType]
 
-  def getParamsList(preparedStatement: PreparedStatement): List[DataType.Name]
+  def getParamsList(preparedStatement: PreparedStatement): List[DataType]
+}
+
+object SessionCollectionResolver {
+  def get(session:Session, name:String):Option[Any] = session.attributes.get(name)
+
+  def getClz[T <: Any](session:Session, name:String):Option[Class[T]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case clz:Class[T]@unchecked => Option(clz)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  // Note the Java return type here.  This function is used to generate objects which will subsequently
+  // be passed to the Java Bindable impl which is expecting Java Iterators.
+  def getIterable[T <: Any](session:Session, name:String):Option[Iterable[T]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case rv:Iterable[T]@unchecked => Option(rv)
+        case rv:lang.Iterable[T]@unchecked => Option(rv.asScala)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  // Note the Java return type here.  This function is used to generate objects which will subsequently
+  // be passed to the Java Bindable impl which is expecting Java Maps.
+  def getMap[K <: Any, V <: Any](session:Session, name:String):Option[Map[K,V]] = {
+    get(session,name).flatMap((sessionVal) => {
+      sessionVal match {
+        case rv:Map[K,V]@unchecked => Option(rv)
+        case rv:util.Map[K,V]@unchecked => Option(rv.asScala.toMap)
+        case _ => Option.empty
+      }
+    })
+  }
+
+  def getIterableClz[T <: Any](session:Session, name:String):Class[_ <: T] = {
```

Comment:
Why are we now storing class information in the Gatling session?
`master` doesn't do that, it looks like it was able to use the setter that doesn't take a class argument:
```java
<T> BoundStatement setSet(int i, Set<T> v)
```
That still exists in 4.x.

Reply:
It does?  SettableByName appears to just have this:

```java
  @NonNull
  @CheckReturnValue
  default <ElementT> SelfT setSet(
      @NonNull String name, @Nullable Set<ElementT> v, @NonNull Class<ElementT> elementsClass) {
    return setSet(firstIndexOf(name), v, elementsClass);
  }
```

SettableById looks to be similar.  Am I missing something?

Reply:
My bad, I should have actually checked 🤦‍♂

That being said, we could easily reimplement it in the plugin code (Java but you get the idea):
```java
  static <SettableT extends SettableByIndex<SettableT>> SettableT setSet(
      SettableT settable, Set<?> set, int index) {
    DataType targetCqlType = settable.getType(index);
    TypeCodec<Object> codec = settable.codecRegistry().codecFor(targetCqlType);
    ByteBuffer encoded = codec.encode(set, settable.protocolVersion());
    return settable.setBytesUnsafe(index, encoded);
  }
```

It's less type-safe (and why it's not in the driver API anymore), but for the way this is going to be used in Gatling tests I think it's good enough. If I understand correctly, the burden of maintaining the "-clz" session parameters falls onto the Gatling user, that's a lot of extra ceremony for little benefit: type mismatches can only be detected at runtime anyway.

Reply:
Ooooh, I like the idea here a great deal.  The "clz" props in the session were added in the absence of a single source of truth for the type information in question but the SettableByName instance certainly could serve in that role.  Very interesting indeed.

One thing I _don't_ like, though, is the explicit encoding + byte-level set here... I'm worried about it's effect on code maintainability since it's not quite as obvious what's going on.  I'm wondering if there's some way to get the Java type from the codec and then still use the member-type aware setSet() method.  You'd wind up doing a double lookup on the codec in that case which isn't... awesome, but maybe not that big a deal.

I want to chew on this idea just a bit more.

Reply:
You could replace the last two lines by:
```java
return settable.set(index, set, codec.getJavaType());
```

Reply:
Had to tweak this just a bit, but I think I have something working now.  codec.getJavaType() returns the type of the collection and we need the individual element type here so there's a bit of extra processing... but it looks like it's gonna work fairly nicely.  I think it is a bit more readable/maintainable than the direct write option.

Reply:
@olim7t when you have time would you mind taking a quick pass over the state of this last change?  I'd like a non-me set of eyes to look at it and confirm it seems sensible